[Part 1] Adding Synthetic NLP Tokenizers, Models, Datasets w/o Integration

moinnadeem opened the pull request:
This PR adds the synthetic NLP tokenizers, models, and datasets. The next PR [Part 2] will take care of integrating these into our models, and updating the actual tests to pass.

ravi-mosaicml commented on composer/models/bert/bert_hparams.py:
```diff
@@ -50,8 +50,10 @@ def initialize_object(self) -> "BERTModel":
                                                                                     **model_hparams)
         else:
+            for k, v in model_hparams.items():
+                assert getattr(config, k) == v
```
is this assert for validation or an invariant? if it's validation, can you make it into a ValueError?

moinnadeem replied:
It's an invariant! The purpose is to ensure that internally casting from the model hparams to an HF config doesn't lose keys. I can add a comment describing why

ravi-mosaicml replied:
Discussed offline, but thoughts about hardcoding the model_hparams in line 55 and 57, since the dict contains one item with a constant key?

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+    return tokenizer
+
+
+class SyntheticHFDataset:
```
Does this need a base class?

moinnadeem replied:
Nope, the base class is `object`

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+            import datasets
+        except ImportError as e:
+            raise ImportError(
+                'Composer was installed without NLP support. To use NLP with Composer, run: `pip install mosaicml[nlp]`.'
```
The import error should have a corresponding statement for anaconda.

moinnadeem replied:
Resolved!

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+    )
+
+
+def generate_synthetic_tokenizer(tokenizer_family: str, dataset=None, vocab_size=256) -> PreTrainedTokenizer:
```
dataset and vocab size are missing type annotations

ravi-mosaicml replied:
also docstring or make private

moinnadeem replied:
Done on both!

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+
+    # flatten dataset to get a simple stream of text for the tokenizer
+    dataset = [dataset[key] for key in dataset.column_names if key != 'idx']
+    dataset = [i for sublist in dataset for i in sublist]
```
can you split this into two statements? double `for in` is hard to follow

moinnadeem replied:
Done!

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+
+
+def generate_bert_tokenizer_params(dataset) -> SyntheticTokenizerParams:
+    try:
```
docstring or make a private method

moinnadeem replied:
Done!

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+
+
+def generate_gpt2_tokenizer_params() -> SyntheticTokenizerParams:
+    try:
```
docstring or make a private method

moinnadeem replied:
Done!

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+                             pad_type_id=0,
+                             pad_token=tokenizer_params.pad_token,
+                             pad_to_multiple_of=8)
+    tokenizer.normalizer = tokenizer_params.normalizer  # type: ignore
```
why the type ignores?

moinnadeem replied:
I was getting errors like: _ error: Cannot assign member "normalizer" for type "Tokenizer"
    Property "normalizer" has no defined setter (reportGeneralTypeIssues)_

This seemed unrelated to our code, but happy to discuss!

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+    tokenizer.train_from_iterator(dataset, trainer=tokenizer_trainer)
+
+    # save the tokenizer config
+    tmp_tokenizer_dir = mkdtemp()
```
use `with tempfile.TemporaryDirectory() as tempdir` instead of `mkdtemp`...tempfile ensures that the directory is removed at the end.

moinnadeem replied:
Done, this is now possible after the refactor! Earlier on, I didn't want the temporary directory being immediately deleted.

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+
+    # generate a synthetic dataset with reasonable defaults is none is provided
+    if dataset is None:
+        num_samples = 100
```
should these be parameters?

moinnadeem replied:
In my opinion, if the user wants to change these parameters, they should pass in a dataset themselves. We could make these into parameters, and pass that into the dataset (to the same effect), but I thought this would enable a short function signature with the same flexibility. In other words, those are just defaults

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+    dataset = [i for sublist in dataset for i in sublist]
+
+    if "bert" in tokenizer_family:
+        tokenizer_params = generate_bert_tokenizer_params(dataset)
```
instead of taking the `tokenizer_family`, could you take the `tokenizer_params`? That way a user could write their own tokenizer params generation function and still use this synthetic dataset. Or, alternatively, they could use the bert / gpt2 functions that you provide. In this case, then `dataset` will need to be a required argument and constructed beforehand.

ravi-mosaicml commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+        json.dump({"pad_token": tokenizer_params.pad_token}, f)
+
+    # instantiate the new tokenizer
+    assert issubclass(tokenizer_params.tokenizer_cls, PreTrainedTokenizer)
```
is this validation or an invariant?

moinnadeem replied:
Validation! I just changed the assert into a ValueError!

moinnadeem replied:
...wow, i sound way too excited about that

ravi-mosaicml reviewed (commented):
Overall LGTM; left some code style comments. CC @abhi-mosaic @anisehsani @growlix for a more NLP-focused review.

hanlint commented on tests/datasets/test_synthetic_data.py:
```diff
@@ -5,7 +5,81 @@
+@pytest.mark.parametrize('chars_per_sample', [128])
+@pytest.mark.parametrize('column_names', [['sentence'], ['sentence1', 'sentence2']])
+@pytest.mark.parametrize('tokenizer_family', ['bert', 'gpt2'])
+def test_synthetic_hf_dataset_creation(num_samples: int, chars_per_sample: int, column_names: list,
```
Recommend breaking up all the asserts here to individual tests, using fixtures for setup, e.g.
```
@pytest.mark.parametrize('num_samples', [50])
@pytest.mark.parametrize('chars_per_sample', [128])
@pytest.mark.parametrize('column_names', [['sentence'], ['sentence1', 'sentence2']])
@pytest.mark.parametrize('tokenizer_family', ['bert', 'gpt2'])
class TestSyntheticNLP:

   @pytest.fixture
    def dataset_generator(...)
         return SyntheticHFDatset(..)
    
    @pytest.fixture
    def dataset(dataset_generator):
        return dataset_generator.generate_dataset()

    def test_generator_sample(dataset_generator, chars_per_sample):
        sample = dataset_generator.generate_sample()
        assert len(sample) == chars_per_sample

    def test_dataset_shape(dataset, ...):
        assert len(dataset) ==
        assert len(datsaet[..) == 
```

And so forth for the individual tokenizer tests, etc. Essentially everywhere you have a comment below should be converted into an isolated test.



```

moinnadeem replied:
Yeah, I agree -- thanks for pointing this out, I didn't know about this!

moinnadeem replied:
Resolving this since it is done

moinnadeem commented:
Thanks for the comments everyone! I'll resolve these shortly

Adding @growlix for an NLP-focused review!

moinnadeem commented on tests/datasets/test_synthetic_data.py:
```diff
@@ -5,7 +5,81 @@
+@pytest.mark.parametrize('vocab_size', [512])
+def test_synthetic_tokenizer_creation(tokenizer_family, vocab_size):
+    try:
+        from transformers import BertTokenizer, GPT2Tokenizer
```
Done!

moinnadeem commented on composer/datasets/synthetic.py:
```diff
@@ -10,6 +18,13 @@
 from composer.utils.string_enum import StringEnum
 
+if TYPE_CHECKING:
+    import tokenizers.decoders as decoders
```
> is it possible to do from tokenizers import normalizers, pre_tokenizers, ... etc? or just do import tokenizers

I slacked you about this! I originally did `import tokenizers`, but PyRight was complaining.

I can refactor this into multiple synthetic files if we think the file is getting too long!

moinnadeem commented on composer/datasets/synthetic.py:
```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+    initial_alphabet = list(set(initial_alphabet))
+
+    return SyntheticTokenizerParams(
+        tokenizer_model=tokenizers_models.WordPiece(unk_token=unk_token),  # type: ignore
```
Without it, I get the following error about default values that haven't been filled in:  `error: Arguments missing for parameters "vocab", "max_input_chars_per_word" (reportGeneralTypeIssues)` from pyright

moinnadeem commented:
Cool, I think this is ready for another review!

@ravi-mosaicml , I addressed all of your feedback! Let me know if I missed anything. I left the conversations unresolved so you can take a look at my comments for each item.

@hanlint I moved things into fixtures, and created `synthetic_lm` and `test_synthetic_lm_data` files in order to declutter stuff. Good call on both of those!

hanlint commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+
+def _generate_bert_tokenizer_params(dataset) -> SyntheticTokenizerParams:
+    try:
+        import tokenizers.decoders as decoders
```
this is pretty verbose and repeated multiple times in this file. i think we need to come up with a general pattern for files that need to be imported, but have these optional dependencies. cc: @ravi-mosaicml or @dblalock for suggestions?

moinnadeem replied:
Strong agree here, I'm pretty interested in fixing this!

dblalock replied:
I don't know what the best way of dealing with this is. I could imagine a few solutions:

1.  We have a `composer.tokenizers` that lazily tries to import hf tokenizers in its getattr.
2.  We have a singleton imported as `composer.tokenizers` that does the same thing.
3.  We have a `composer.tokenizers` that tries to import hf tokenizers, but instead creates a Mock if that fails. The Mock throws the informative ImportError below whenever it's called.

3 seems the least bad, but none of these seem ideal.

moinnadeem replied:
Discussed offline: out of scope for this PR

hanlint commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+                             pad_type_id=0,
+                             pad_token=tokenizer_params.pad_token,
+                             pad_to_multiple_of=8)
+    tokenizer.normalizer = tokenizer_params.normalizer  # type: ignore
+    tokenizer.pre_tokenizer = tokenizer_params.pre_tokenizer  # type: ignore
+    tokenizer.decoder = tokenizer_params.decoder  # type: ignore
```
type ignore?

hanlint replied:
Is this because your are assigning attributes to tokenizer that are not in its public API?

moinnadeem replied:
Correct! See this: https://github.com/mosaicml/composer/pull/650#discussion_r821361537

Is there a better way to do this? Happy to implement it if so

ravi-mosaicml replied:
It seems like that the underlying `.pyi` file has incorrect annotations. Let's paste in the error message here.

moinnadeem replied:
Done!

hanlint commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+    if not issubclass(tokenizer_params.tokenizer_cls, PreTrainedTokenizer):
+        raise ValueError(f"{tokenizer_params.tokenizer_cls} should sub-class transformers.PreTrainedTokenizer.")
+    tokenizer = tokenizer_params.tokenizer_cls.from_pretrained(tmp_tokenizer_dir)
+
```
should the tempdir be cleaned up here? if so, perhaps instead should use the `TemporaryFolder` context instead of the low level `mkdtemp`

moinnadeem replied:
The temporary directory should not be cleaned up, because the tokenizer folder that is created needs to exist for the model to access the tokenizer as well.

This might imply a potential race condition if the temporary directory is cleaned up by the OS before the model gets instantiated? Is that a concern?

ravi-mosaicml replied:
Discussed offline; the folder will be passed in as an argument into the function.

moinnadeem replied:
Done!

hanlint commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+            import datasets
+        except ImportError as e:
+            raise ImportError(
+                textwrap.dedent("""\
```
For DRY, we probably want to store the message in a constant up front, e.g.
```
_IMPORT_MESSAGE = textwrap.dedent(...)
```
and then here just do:
```
raise ImportError(_IMPORT_MESSAGE) from e
```
```

moinnadeem replied:
Agreed, happy to implement this.

ravi-mosaicml replied:
Use #795

moinnadeem replied:
Done!

hanlint reviewed (commented):
LGTM, with a few comments, with one major comment on establishing a good pattern for optional imports that we should raise for a braoder discussion.

[nit] Not critical, but could benefit from a bit more tweaking on the pytest structure.

moinnadeem commented:
> [nit] Not critical, but could benefit from a bit more tweaking on the pytest structure.

Can you provide any suggestions? I might be interested in implementing it, but at the very least, I'd be interested in learning what I could have done better.

growlix commented on tests/datasets/test_synthetic_lm_data.py:
```diff
@@ -0,0 +1,117 @@
+from itertools import product
```
This isn't standard practice, but I wonder if we should start annotating our tests a bit more. Their intent can often be inferred from the code, but it might make things easier for everyone if their intent was clearly stated by the author.

growlix reviewed (approved):
Minor comments, mostly regarding documentation. LGTM.

growlix commented on tests/datasets/test_synthetic_lm_data.py:
```diff
@@ -0,0 +1,117 @@
+from itertools import product
+
```
Does it make sense to test whether the `[sep]`, `[mask]`, and `[cls]` tokens are present as expected? More generally, are there any assumptions that algorithms make about the structure of the input data, and do those need to be checked here?

moinnadeem replied:
Yes! That is what `test_tokenizer_specific_properties` does!

growlix reviewed (approved):
Added a comment about tests

ravi-mosaicml commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+    )
+
+
+def generate_synthetic_tokenizer(tokenizer_family: str,
```
Add in the tokenizer directory as an argument here, instead of creating the tempdir in the function below

moinnadeem replied:
Done!

ravi-mosaicml commented on tests/datasets/test_synthetic_lm_data.py:
```diff
@@ -0,0 +1,117 @@
+from composer.datasets.synthetic_lm import SyntheticHFDataset, generate_synthetic_tokenizer
+
+
+def generate_parameter_configs(keys, num_replicas=1):
```
Add in a comment on what is happening here...perhaps with an example output and how this is then used by the `dataset_generator` fixture below.

moinnadeem replied:
Done!

ravi-mosaicml commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+        return hf_synthetic_dataset
+
+    def generate_sample(self):
+        MIN_WORD_LENGTH = 3
```
can you make these init args with default values?

moinnadeem replied:
Done!

ravi-mosaicml commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+
+class SyntheticHFDataset:
+    """Creates a synthetic HF dataset and passes it to the preprocessing scripts."""
+
```
Discussed offline to refactor this into a function (since it's really a builder, not a dataset), and then make `generate_sample` a private, module-level function.

moinnadeem replied:
Done!

moinnadeem commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
```
Done! Good catch!

moinnadeem commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+    import tokenizers.pre_tokenizers as pre_tokenizers
+    from datasets import Dataset
+    from transformers import PreTrainedTokenizer
+
```
Not down to be president, but done regardless!

moinnadeem commented on composer/datasets/synthetic_lm.py:
```diff
@@ -0,0 +1,231 @@
+    if "bert" in tokenizer_family:
+        tokenizer_params = _generate_bert_tokenizer_params(flattened_dataset)
+    elif "gpt2" in tokenizer_family:
+        tokenizer_params = _generate_gpt2_tokenizer_params()
```
Ah, good question! This is a difference in WordPiece vs BPE: WordPiece starts with an initial vocabulary (which is drawn from the dataset), and gradually merges characters together into tokens. BPE starts with the text, and gradually breaks text down into tokens. So one requires the initial vocabulary, the other does not. Does this make sense?

moinnadeem commented:
@ravi-mosaicml @hanlint Cool, addressed all comments!

@hanlint btw, I resolved the `mkdtemp` thing by using PyTest fixtures to pass in a temporary directory, which can exist for the model tests as well.

hanlint reviewed (approved):
LGTM, will merge after tests pass.
